src/test/app/Clawback_test.cpp

```diff
@@ -0,0 +1,971 @@
+            env.require(balance(cindy, alice["USD"](1000)));
+
+            // alice claws back 300 USD from cindy
+            env(claw(alice, cindy["USD"](300)));
+            env.close();
+
+            // bob's balance is unchanged
+            env.require(balance(alice, bob["USD"](-100)));
+            env.require(balance(bob, alice["USD"](100)));
+
+            // cindy's balance is reduced
+            env.require(balance(alice, cindy["USD"](-700)));
+            env.require(balance(cindy, alice["USD"](700)));
+        }
+    }
+
+    void
+    testBidirectionalLine(FeatureBitset features)
+    {
+        testcase("Bidirectional line");
+        using namespace test::jtx;
+
+        // Test when both alice and bob issues USD to each other.
+        // This scenario creates only one trustline.
+        // In this case, both alice and bob can be seen as the "issuer"
+        // and they can send however many USDs to each other.
+        // We test that only the person who has a negative balance from their
+        // perspective is allowed to clawback
+        Env env(*this, features);
+
+        Account alice{"alice"};
+        Account bob{"bob"};
+
+        env.fund(XRP(1000), alice, bob);
+        env.close();
+
+        // alice sets asfAllowClawback
+        env(fset(alice, asfAllowClawback));
+        env.close();
+        env.require(flags(alice, asfAllowClawback));
+
+        // bob sets asfAllowClawback
+        env(fset(bob, asfAllowClawback));
+        env.close();
+        env.require(flags(bob, asfAllowClawback));
+
+        // alice issues 1000 USD to bob
+        env.trust(alice["USD"](1000), bob);
+        env(pay(alice, bob, alice["USD"](1000)));
+        env.close();
+
+        BEAST_EXPECT(ownerCount(env, alice) == 0);
+        BEAST_EXPECT(ownerCount(env, bob) == 1);
+
+        // bob is the holder, and alice is the issuer
+        env.require(balance(bob, alice["USD"](1000)));
+        env.require(balance(alice, bob["USD"](-1000)));
+
+        // bob issues 1500 USD to alice
+        env.trust(bob["USD"](1500), alice);
+        env(pay(bob, alice, bob["USD"](1500)));
+        env.close();
+
+        BEAST_EXPECT(ownerCount(env, alice) == 1);
+        BEAST_EXPECT(ownerCount(env, bob) == 1);
+
+        // bob has negative 500 USD because bob issued 500 USD more than alice
+        // bob can now been seen as the issuer, while alice is the holder
+        env.require(balance(bob, alice["USD"](-500)));
+        env.require(balance(alice, bob["USD"](500)));
+
+        // At this point, both alice and bob are the issuers of USD
+        // and can send USD to each other through one trustline
+
+        // alice fails to clawback. Even though she is also an issuer,
+        // the trustline balance is positive from her perspective
+        env(claw(alice, bob["USD"](200)), ter(tecNO_PERMISSION));
+        env.close();
+
+        // bob is able to successfully clawback from alice because
```

Comment:
The party with a negative balance can unilaterally clawback from the counter-party of the trustline right?

Are there real-world analogues to this type of behavior? It feels quite arbitrary (although I'm aware that both the parties need to set the `asfClawback` flag and enable this particular amendment)